_posts/2022-10-27-using-kani-with-the-bolero-property-testing-framework.md

```diff
@@ -0,0 +1,311 @@
+# snip ...
+#3773   REDUCE cov: 135 ft: 138 corp: 18/282b lim: 52 exec/s: 0 rss: 43Mb L: 8/32 MS: 3 ChangeByte-ChangeBinInt-EraseBytes-
+#262144 pulse  cov: 135 ft: 138 corp: 18/282b lim: 2620 exec/s: 131072 rss: 68Mb
+#524288 pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 131072 rss: 93Mb
+#1048576        pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 116508 rss: 142Mb
+```
+
+The fuzzer will run for a while without finding any failures. We can now run Kani to *prove* the assertion in our harness:
+
+```
+$ cargo bolero test check_rev --engine kani
+# snip ...
+SUMMARY:
+ ** 0 of 338 failed (1 unreachable)
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.90419155s
+```
+
+and voila! Kani managed to prove that there none of the checks can fail for any values of the inputs.
+
+## If we have proof capability, do we still need fuzzing?
+
+One question that might arise is: is there any point in using fuzzing alongside with Kani?
+In other words, if we can prove a harness for all inputs with Kani, would we still want to use fuzzing?
+The answer is yes, because fuzzing and Kani provide complementary benefits.
+Kani verifies a Rust program by symbolically analyzing its code.
+This allows Kani to make mathematical statements about the expected semantics of the Rust code being verified.
+Kani allows you to prove that, **for all possible inputs**, the code under verification follows its specification, **assuming everything else functions correctly**.  
+Fuzzing concretely executes the program under test.
+This gives you end-to-end confidence that **for the set of inputs generated by the fuzzer**, the code under verification follows its specification, **under real-world conditions**.
+Fuzzing and Kani fit together to give more assurance than either provides on its own.
+
+## How is Kani integrated in Bolero?
+
+For those that are curious, let's briefly discuss how the `kani` engine is implemented.
+At a high-level, Bolero redirects value generation to `kani::any()`, and calls `cargo kani` under the hood when `cargo bolero test` is invoked with the `kani` engine.
+Also, the `for_each` function only iterates once in the `kani` mode, since `kani::any()` covers all possible values of an input.
+
+Some of the generation capabilities of Bolero require a little more than just calling `kani::any()`.
+For instance, Bolero supports generating values in a restricted range `[min, max)`, e.g.
+
+```rust
+bolero::check!().with_generator(5..37).for_each( // snip... );
+```
+
+To support those in the Kani mode, Bolero leverages assumptions (`[kani::assume](https://model-checking.github.io/kani/tutorial-first-steps.html#assertions-assumptions-and-harnesses)`) under the hood:
+
+```rust
+let value = kani::any();
+kani::assume((min..max).contains(&value));
+value
+```
+
+To get started, check out the Bolero and Kani tutorials at the following links:
+ * [Bolero tutorial](https://camshaft.github.io/bolero/tutorials/fibonacci.html)
```

Comment:
We're missing a tutorial for kani / bolero integration. :(

Reply:
Yes, we should hopefully add one shortly. Created https://github.com/model-checking/kani/issues/1807 to track this.